provide profile as part of S3ClientConfig

TalIfargan opened the pull request:
## Description
Everything is ready to support custom S3 user profiles. We only need to allow passing it through the `S3ClientConfig` to the `MountpointS3Client`.

## Additional context
The rust implementation is already ready to accept a custom user provided profile - we only need to reveal it to the user as part of the `S3ClientConfig` configuration.


- [X] I have updated the CHANGELOG or README if appropriate


## Testing
I have tested this PR with a custom profile and everything worked as expected. This PR is really a 4 line change that could make a nice impact for specific use cases.

--------
By submitting this pull request, I confirm that my contribution is made under the terms of BSD 3-Clause License and I agree to the terms of the [LICENSE](https://github.com/awslabs/s3-connector-for-pytorch/blob/main/LICENSE).

IsaevIlya commented on s3torchconnector/src/s3torchconnector/_s3client/s3client_config.py:
```diff
@@ -15,12 +15,15 @@ class S3ClientConfig:
     force_path_style(bool): forceful path style addressing for S3 client.
-    max_attempts(int): amount of retry attempts for retrieable errors
+    max_attempts(int): amount of retry attempts for retrieable errors.
+    config(str): Profile name to use for S3 authentication.
```
Should it be `profile(str)`?

IsaevIlya commented on s3torchconnector/src/s3torchconnector/_s3client/s3client_config.py:
```diff
@@ -15,12 +15,15 @@ class S3ClientConfig:
     unsigned: bool = False
     force_path_style: bool = False
     max_attempts: int = 10
+    config: str = None
```
Should it be `profile: str`?

IsaevIlya reviewed (changes requested):
Hey @TalIfargan,
Thank you for your valuable contribution! The changes look promising overall. I've added two minor comments for your review. We'll need some time to properly evaluate and test this functionality, after which we'll provide further feedback.

TalIfargan commented:
> Hey @TalIfargan, Thank you for your valuable contribution! The changes look promising overall. I've added two minor comments for your review. We'll need some time to properly evaluate and test this functionality, after which we'll provide further feedback.

Hi @IsaevIlya, thanks for your swift review.

I have addressed these two minor comments.
I think that after you are able to test this properly we might also update the README to include this is an option for authentication.

IsaevIlya commented:
>I think that after you are able to test this properly we might also update the README to include this is as an option for authentication.
 
Good catch. Would you be able to update the documentation while we complete the testing?

TalIfargan commented:
> > I think that after you are able to test this properly we might also update the README to include this is as an option for authentication.
> 
> Good catch. Would you be able to update the documentation while we complete the testing?

Sure.

TalIfargan commented:
Any news about this?

IsaevIlya commented:
Hey @TalIfargan ,
Thank you for your patience. We've made progress by updating our infrastructure to support this feature. Our next step is to integrate it into our CI/CD pipeline. I'll update you, as soon as it will be done.

IsaevIlya commented:
Hey @TalIfargan ,
We have extended our CI/CD pipeline to support authentication using profile. Could you please add corresponding test to [test_s3_client.py](https://github.com/awslabs/s3-connector-for-pytorch/blob/main/s3torchconnector/tst/e2e/test_s3_client.py). For reference, you can follow a recently added test [test_access_objects_with_profile](https://github.com/awslabs/s3-connector-for-pytorch/commit/56d632d4df514dd9f84a7eedd7274c940c0d4016#diff-ee00ca5b6583e2b300fbea4330d04c1be0e2771521ec4da7ca165a8f30f804a8R512-R552). This new test should verify the profile-based authentication functionality at the S3Client level, similar to how we're testing the underlying implementation.

TalIfargan commented:
> Hey @TalIfargan , We have extended our CI/CD pipeline to support authentication using profile. Could you please add corresponding test to [test_s3_client.py](https://github.com/awslabs/s3-connector-for-pytorch/blob/main/s3torchconnector/tst/e2e/test_s3_client.py). For reference, you can follow a recently added test [test_access_objects_with_profile](https://github.com/awslabs/s3-connector-for-pytorch/commit/56d632d4df514dd9f84a7eedd7274c940c0d4016#diff-ee00ca5b6583e2b300fbea4330d04c1be0e2771521ec4da7ca165a8f30f804a8R512-R552). This new test should verify the profile-based authentication functionality at the S3Client level, similar to how we're testing the underlying implementation.

Hi, @IsaevIlya,

I have added the requested test to [test_s3_client.py](https://github.com/awslabs/s3-connector-for-pytorch/blob/main/s3torchconnector/tst/e2e/test_s3_client.py), following the reference test you have mentioned. Thanks for putting your effort into this.

IsaevIlya reviewed (changes requested):
Oh, there are linter errors. Could you please run `black` to reformat `test_s3_client_config.py`?

TalIfargan commented:
updated

IsaevIlya reviewed (approved):
Looks good, thank you!
